lib/ruby_lsp/requests/hover.rb

```diff
@@ -0,0 +1,109 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "ruby_lsp/requests/support/rails_document_client"
+
+module RubyLsp
+  module Requests
+    # ![Hover demo](../../misc/rails_document_link_hover.gif)
+    #
+    # The [hover request](https://microsoft.github.io/language-server-protocol/specification#textDocument_hover)
+    # renders a clickable link to the code's official documentation.
+    # It currently only supports Rails' documentation: when hovering over Rails DSLs/constants under certain paths,
+    # like `before_save :callback` in `models/post.rb`, it generates a link to `before_save`'s API documentation.
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Post < ApplicationRecord
+    #   before_save :do_something # when hovering on before_save, the link will be rendered
+    # end
+    # ```
+    class Hover < BaseRequest
+      extend T::Sig
+
+      sig { params(document: Document, uri: String, position: Document::PositionShape).void }
+      def initialize(document, uri, position)
+        super(document)
+
+        @file_dir = T.let(Pathname.new(uri).dirname.to_s, String)
+        @hover = T.let(nil, T.nilable(LanguageServer::Protocol::Interface::Hover))
+        position = Document::Scanner.new(document.source).find_position(position)
+        @target = T.let(find(T.must(@document.tree), position), T.nilable(SyntaxTree::Node))
+      end
+
+      sig { override.returns(T.nilable(LanguageServer::Protocol::Interface::Hover)) }
+      def run
+        return if @target.nil?
+
+        visit(@document.tree)
+        @hover
+      end
+
+      sig { override.params(node: SyntaxTree::Command).void }
+      def visit_command(node)
+        return unless node == @target
+
+        message = node.message
+        generate_rails_document_link_hover(message.value, message)
+      end
+
+      sig { override.params(node: SyntaxTree::FCall).void }
+      def visit_fcall(node)
+        return unless node == @target
+
+        message = node.value
+        generate_rails_document_link_hover(message.value, message)
+      end
+
+      sig { override.params(node: SyntaxTree::ConstPathRef).void }
+      def visit_const_path_ref(node)
```

Comment:
I understand we need the `find` technique for document highlight, but I'm not sure it's necessary for this one.

Instead of inspecting the AST twice (once during `find` and another one during `visit`), couldn't we do something like this?

```ruby
def visit_fcall(node)
  # We found an fcall, which is one of the nodes we're interested in. As long as its location covers the position,
  # we should be able to push the hover, no?
  location = node.location
  return unless (location.start_line - 1..location.end_line - 1).cover?(@position[:line]) && (location.start_column..location.end_column).cover?(@position[:character])
end
```
We can even put it in the `BaseRequest` given that it's common for requests to check if a node covers the position. E.g.:
```ruby
def visit_fcall(node)
  return unless node_covers_position?(node, @position)
end
```